### ISSUE TITLE
contains type variables that cannot be generalized

### ISSUE THREAD

**Issue by Ulrar** [andreas/ocaml-graphql-server#124]
Hi,

When I try to declare my types in their own files, I get an error, for example if I have a file with

```ocaml
let projectGraphqlObj =
  let open Schema in
  obj "project"
    ~doc:"project"
    ~fields:
      (fun _ ->
        [ field "name"
            ~doc:"Project's name"
            ~typ:(non_null string)
            ~args:Arg.[]
            ~resolve:(fun _ project -> project.name)
        ]
      )
```
If the actual Schema that uses this isn't in the same file, I get this when compiling : 
```
Error: The type of this expression,
       ('_weak1, project option) Graphql.Schema.typ,
       contains type variables that cannot be generalized
```

My best guess is that it must be coming from the ignored ctx parameters to the functions, whatever it's for, but named or not I still get the error of course. Shoving everything into one file "solves" it, but that file is getting ridiculously big now, I'd like to be able to split it properly if at all possible.
Thanks !

**Comment by anmonteiro** [comment]
You're right that it's coming from the `ctx' parameters. I think if you annotate them the error will be gone.

Example:

```ocaml
~resolve:(fun (ctx: my_ctx_type) project -> project.name)
```

**Comment by Ulrar** [comment]
Ah, indeed. Using (ctx : unit) instead of _ seems to do the trick, thanks

**Comment by anmonteiro** [comment]
If `ctx` is of `unit` type, then you can just destructure it: `fun () project -> ...`
